### PR TITLE
Interop with Akka Persistence plugins

### PR DESCRIPTION
**Pull request by dfakhritdinov**
This PR is second in the sequence of depended PRs:
1. [New persistence api module](https://github.com/evolution-gaming/akka-effect/pull/273)
2. [Interop w/ Akka Persistence plugins](https://github.com/evolution-gaming/akka-effect/pull/274), this one
3. [New persistent actor impl](https://github.com/evolution-gaming/akka-effect/pull/275)

The PR contains interoperability between Akka Persistence plugins and persistence API defined in `persistence-api` module, making all set of available Akka Persistence plugins valid implementations of new persistence API introduced in previous PR. 

Akka Persistence plugins can be accessed as actors via restricted to `akka.persistence` package API thus interop implementation located in the same package. Snapshot plugin interop implemented using "ask pattern" due to its simplicity - each operation represented by one command to the plugin actor and one reply command. Journal plugin interop implemented using `LocalActorRef`, inspired by "ask pattern" but supporting multi-command reply.

The reasoning behind `LocalActorRef` is in avoiding overhead of having new actor (per request) and avoiding state management in one long-leaving actor (that can handle all those commands to snaphotter or journaller actors). Actors are pretty cheap but not fully free and IMO creating new one each time on persisting events will be too much. On other hand managing concurrent requests within one long-living actor (lifespan should be equal to main actor) also better to be avoided cos concurrent requests must be supported by the persistent plugin anyway and having same logic twice not perfect solution. 

As an alternative I'm proposing using `LocalActorRef` - lightweight `ActorRef` implementation allows aggregating result from multiple responses. Its functionality is limited and should not be used outside of `akka-effect` thus it marked as `private[akka.persistence]`

### CODE REVIEW

File: persistence/src/main/scala/akka/persistence/LocalActorRef.scala [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+    val F = Temporal[F]
+
+    case class State(state: S, updated: Instant)
+
+    def timeoutException = new TimeoutException(s"no messages received during period of $timeout")
+
+    for {
+      now   <- F.realTimeInstant
+      state <- SerialRef.of[F, State](State(initial, now))
```
Comment (t3hnar) [inline persistence/src/main/scala/akka/persistence/LocalActorRef.scala:73]: why serial here?

File: persistence/src/main/scala/akka/persistence/LocalActorRef.scala [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+      }
+    } yield new LocalActorRef[F, R] {
+
+      private def done(e: Either[Throwable, R]) =
+        for {
+          _ <- defer.complete(e)
+          _ <- fiber.cancel
+        } yield {}
+
+      override def ref: ActorRef = new MinimalActorRef {
```
Comment (t3hnar) [inline persistence/src/main/scala/akka/persistence/LocalActorRef.scala:96]: why can't your actor just be `(f: Any => F[Unit]) => ActorRef` ?

**Review by t3hnar — CHANGES_REQUESTED** [review]

File: persistence/src/main/scala/akka/persistence/LocalActorRef.scala (hunk 1 above)
Comment (dfakhritdinov) [reply persistence/src/main/scala/akka/persistence/LocalActorRef.scala:73]: Serial is safe protection from calling the ref concurrently: ie from some kind of callback function within persistence implementation. Akka actor ref naturally support concurrent calls thus my fake ref should support them as well

File: persistence/src/main/scala/akka/persistence/LocalActorRef.scala (hunk 2 above)
Comment (dfakhritdinov) [reply persistence/src/main/scala/akka/persistence/LocalActorRef.scala:96]: `ActorRef` trait is pretty huge with lots of methods to be implemented, I prefer to delegate it to `MinimalActorRef`

Comment (t3hnar) [reply persistence/src/main/scala/akka/persistence/LocalActorRef.scala:96]: I was more talking about that you logic of state/io composition all mixed inside of `MinimalActorRef` impl, where you basically just need to call IO.
This is not a strong opinion

Comment (dfakhritdinov) [reply persistence/src/main/scala/akka/persistence/LocalActorRef.scala:96]: Ah, got your point. Yes, I can easily extract the logic into separate function, but I do not see much benefits of it.

On ref creation I defined few variables:
```
val state: SerialRef[F,State] = ???
val defer: Deferred[F,Either[Throwable,R]] = ???
val fiber: Fiber[F,Throwable,Unit] = ???
```

and all three of them used in _the function_ that now is defined in `override def !(m: M)(implicit sender: ActorRef): Unit`. I tend to keep this coupled and incapsulated inside `LocalActorRef` implementation due to code complexity and no much benefits of exposing it outside.

Taking into account my previous statement I can propose somethings like:
```
      private def update(m: M): F[Unit] =
        state
          .update { s =>
            if (receive.isDefinedAt(s.state -> m)) {

              for {
                t <- Temporal[F].realTimeInstant
                r <- receive(s.state -> m)
                s <- r match {
                  case Left(s)  => State(s, t).pure[F]
                  case Right(r) => done(r.asRight).as(s)
                }
              } yield s

            } else {
              s.pure[F]
            }
          }
          .handleErrorWith { e =>
            done(e.asLeft).void
          }

        override def !(m: M)(implicit sender: ActorRef): Unit =
          update(m).toTry.get
```

but IMO its not much difference from how its done now

**Pull request closed without merging**